server.c

```diff
@@ -163,36 +167,109 @@ void *ClientHandler(void *cli)
-{
-	printf("\n");
-	char *file_name = strchr(request, ' ');
-	printf("%s", file_name);
-	return file_name + 1;
-}
+		//obsluga komendy LIST
+		if (strcmp(request_token, "LIST") == 0)
+		{
+			if (client.data_socket > 0)
+			{
+				sendResponse(client.socket, "150 Here comes the directory listing. \r\n");
+				request_token = strtok(NULL, " \r\n");
+				if (request_token == NULL)
+				{
+					char cwd[256];
+					getcwd(cwd, sizeof(cwd));
+					request_token = &cwd[0];
+				}
+				DIR *dir = opendir(request_token);
+				char path[256];
+				struct stat buf;
+				struct tm *t;
+				char timebuff[80];
+				time_t time;
+				if (dir != NULL)
+				{
+					struct dirent *file;
+					while ((file = readdir(dir)))
+					{
+						if ((strcmp(file->d_name, ".") != 0) && (strcmp(file->d_name, "..") != 0))
+						{
+							strcpy(path, request_token);
+							strcat(path, "/");
+							strcat(path, file->d_name);
+							if (stat(path, &buf) != -1)
+							{
+								time = buf.st_mtime;
+								t = localtime(&time);
+								strftime(timebuff, 80, "%b %d %H:%M", t);
+								sprintf(client_response, "%c%s %5d %4d %4d %8d %s %s \r\n",
+										file->d_type == DT_DIR ? 'd' : '-',
+										S_ISDIR(buf.st_mode) ? "rwxr-xr-x" : "rw-r--r--",
+										buf.st_nlink,
+										buf.st_uid,
+										buf.st_gid,
+										buf.st_size,
+										timebuff,
+										file->d_name);
+								sendResponse(client.data_socket, client_response);
+							}
+						}
+					}
+					closedir(dir);
+					sendResponse(client.socket, "226 Listing completed. \r\n");
+					client.data_socket = -1;
+					close(client.data_socket);
+					printf("Listing completed.\n");
+				}
+				else
+				{
+					sendResponse(client.socket, "550 Failed to open directory. \r\n");
+				}
+			}
+			else
+			{
+				sendResponse(client.socket, "425 Use PORT or PASV first. \r\n");
+			}
+		}
 
-// te funkcja jest stara i moze nie miec sensu
-bool SendFileOverSocket(int socket_desc, char *file_name)
-{
+		if (strcmp(request_token, "RETR") == 0)
+		{
+			if (client.data_socket > 0)
+			{
+				request_token = strtok(NULL, " \r\n");
+				int size;
+				char buf[BUFSIZ] = {0};
+				FILE *file = fopen(request_token, "r");
+				if (file == NULL)
+				{
+					sendResponse(client.socket, "550 Requested action not taken. File unavailable.\n");
+				}
 
-	struct stat obj;
-	int file_desc,
-		file_size;
+				else
+				{
+					sendResponse(client.socket, "150 Here comes the file retreiving. \r\n");
+					while ((size = fread(buf, sizeof(char), BUFSIZ, file)) > 0)
+					{
+						buf[size] = '\0';
+						sendResponse(client.data_socket, buf);
+					}
 
-	stat(file_name, &obj);
-	file_desc = open(file_name, O_RDONLY);
-	file_size = obj.st_size;
-	send(socket_desc, &file_size, sizeof(int), 0);
-	sendfile(socket_desc, file_desc, NULL, file_size);
+					sendResponse(client.socket, "226 Retreiving completed. \r\n");
+					fclose(file);
+					client.data_socket = -1;
+					close(client.data_socket);
```

Comment:
to samo co w LIST, przez to klient `bin/ftp` się blokuje i nie kończy procedury bo wydocznie czeka na zamkniecie połączenia.